proposals/0000-local-import.rst

```diff
@@ -0,0 +1,225 @@
+
+Furthermore, it encourages the use of qualified imports, as those can be locally “de-qualified” in order to improve readability.
+
+Costs and Drawbacks
+-------------------
+These changes should be easy to grasp by beginners. Moreover, they could greatly improve learnability of Haskell libraries, by expliciting the module each function comes from in examples and tutorials.
+
+Some existing tooling (e.g., ``snack``) assume that imports are only found at the toplevel and might be broken by this change. Perhaps more importantly, this change would make it harder for IDE-like tools such as ``hie`` to determine the set of valid completions ; such tools would need to be made context-sensitive, like OCaml's merlin.
+
+This change would make it harder to determine at first glance inter-dependencies between modules.
+
+Finally, some library writers might choose to design their library around this extension. Using such libraries without this extension enabled might be inconvenient, which could be perceived as a drawback by some users.
+
+Alternatives
+------------
+To our knowledge, there is no other language feature or extension providing similar benefits.
+
+It might be valuable to also allow some form of typelevel local import ; the shortcut syntax, in particular, could be used to simplify type signatures.
+
+The syntactic shortcut syntax is orthogonal to the rest of the proposal and could be entirely removed. The same functionality could potentially be achieved with QuasiQuoters.
+
+As an extension to the proposed behavior, local imports could be used to shadow globally-defined symbols. As an example, the ``blaze-html`` library provides symbols for ``head``, ``div`` and ``id`` ; for this reason,  the relevant modules are frequently imported qualified, or those symbols are explicitly hidden with ``-XNoImplicitPrelude`` and an explicit import. This is necessary for preventing uses of those symbols to be reported as ambiguous by the compiler. Without type-driven disambiguation, this is the only sane behavior in current Haskell, which only allows a single, unordered list of module imports ; however, local imports could be seen as defining nested scopes, such that:
+::
+
+  {-# LANGUAGE OverloadedStrings #-}
+  import Text.Blaze.Html4.Strict as Blaze
+  import Text.Blaze.Html4.Strict.Attributes as Blaze
+
+  markup :: Html
+  markup = head $ div ! id "foo"
+    where import Blaze
+compiles without error. Similarly, in the following example:
+::
+
+  import Foo as A
+
+  main :: IO ()
+  main = do
+    import Bar as A
+    A.x
+If ``x`` is defined in both ``Foo`` and ``Bar``, the import from ``Bar`` could take precedence over the one from ``Foo``. Finally, DSLs could benefit from this change to override arithmetic operators without implementing bogus ``Num`` instances.
```

Comment:
« bogus » is a loaded term, you probably want to avoid it.

Also, readers don't necessarily share where you're coming from here. Explain what such an instance looks like (with some undefined symbols). And how your idea of hierarchical shadowing could solve this problem.